libs/networking/Http_mock_client.ml

```diff
@@ -126,3 +136,146 @@ let with_testing_client make_fn test_fn () =
+      let i = String.index s '\n' in
+      let before, after = (Str.first_chars s i, Str.string_after s (i + 1)) in
+      go after (before :: acc)
+    else (acc, s)
+  in
+  let acc, rem = go s [] in
+  (acc |> List.rev |> p, rem)
+
+let parse_first_line_req s =
+  let tokens = String.split_on_char ' ' s in
+  match tokens with
+  | [ meth; resource; ver ] ->
+      ( Cohttp.Code.method_of_string meth,
+        resource,
+        Cohttp.Code.version_of_string ver )
+  | _ -> failwith "Invalid format"
+
+let parse_first_line_resp s =
+  let tokens = String.split_on_char ' ' s in
+  match tokens with
+  | version :: status_code :: status_text ->
+      ( Cohttp.Code.version_of_string version,
+        Cohttp.Code.status_of_code (int_of_string status_code),
+        String.concat " " status_text )
+  | _ -> failwith "Invalid format"
+
+let trim_front s =
+  let len = String.length s in
+  let rec go s i =
+    if len < i then len
+    else if List.mem (String.get s i) [ '\t'; '\n'; '\r'; ' '; '\x0C' ] then
+      go s (i + 1)
+    else i
+  in
+  Str.string_after s (go s 0)
+
+let parse_headers headers =
+  headers
+  |> Common.map (fun header ->
+         let i = String.index header ':' in
+         let before, after =
+           ( Str.first_chars header i,
+             Str.string_after header (i + 1) |> trim_front )
+         in
+         (before, after))
+  |> Cohttp.Header.of_list
+
+let take_while p xs =
+  let rec go xs acc =
+    match xs with
+    | [] -> (List.rev acc, None, [])
+    | x :: xs -> if p x then go xs (x :: acc) else (List.rev acc, Some x, xs)
+  in
+  go xs []
+
+let parse_req =
+  strip_and_parse "> " (fun lines ->
+      match lines with
+      | [] -> failwith "Invalid format"
+      | first_line :: lines ->
+          let meth, resource, version = parse_first_line_req first_line in
+          let headers, _, body =
+            take_while (fun s -> String.length s <> 0) lines
+          in
+          let headers = parse_headers headers in
+          let body = String.concat "\n" body |> Body.of_string in
+          ( {
+              Cohttp.Request.meth;
+              resource;
+              version;
+              headers;
+              scheme = None;
+              encoding = Body.transfer_encoding body;
+            },
+            body ))
+
+let parse_resp =
+  strip_and_parse "< " (fun lines ->
+      match lines with
+      | [] -> failwith "Invalid format"
+      | first_line :: lines ->
+          let version, status, _ = parse_first_line_resp first_line in
+          let headers, _, body =
+            take_while (fun s -> String.length s <> 0) lines
+          in
+          let headers = parse_headers headers in
+          let body = String.concat "\n" body |> Body.of_string in
+          ( {
+              Cohttp.Response.version;
+              headers;
+              status;
+              (* Not sure what this field is for. Not documented *)
+              flush = false;
```

Comment:
Probably that IO should be flushed when writing this body vs. not flushing and allowing chunking/streaming. Most likely want this true, but idk if it'll actually change anything since this isn't ever going over the network

Reply:
I don't think its buffered at all based on my testing, but I can make it true. Cohttp doesn't document the actual use.